src/fivetran-destination/src/destination/dml.rs

```diff
@@ -231,120 +231,83 @@ async fn load_file(file_config: &FileConfig, path: &str) -> Result<AsyncFileRead
-    // at another, rather than presenting as a single update at a single
-    // timestamp.
+    // Copy into a temporary table, which we then merge into the destination.
+    let (qualified_temp_table_name, columns) = create_temporary_table(table, client).await?;
+    let row_count = copy_files(
+        file_config,
+        replace_files,
+        client,
+        &qualified_temp_table_name,
+    )
+    .await
+    .context("replace_files")?;
 
-    let mut key = vec![];
-    let mut delete_stmt = format!(
-        "DELETE FROM {}.{} WHERE ",
-        escape::escape_identifier(schema),
-        escape::escape_identifier(&table.name),
-    );
-    let mut p = 1;
-    for (i, c) in table.columns.iter().enumerate() {
-        if c.primary_key {
-            key.push(i);
-            if p > 1 {
-                delete_stmt += " AND "
-            }
-            delete_stmt += &format!("{} = ${p}", escape::escape_identifier(&c.name));
-            p += 1;
-        }
+    // Bail early if there isn't any work to do.
+    if row_count == 0 {
+        return Ok(());
     }
 
-    let insert_stmt = format!(
-        "INSERT INTO {}.{} ({}) VALUES ({})",
+    let qualified_table_name = format!(
+        "{}.{}",
         escape::escape_identifier(schema),
-        escape::escape_identifier(&table.name),
-        table
-            .columns
-            .iter()
-            .map(|c| escape::escape_identifier(&c.name))
-            .join(","),
-        (1..=table.columns.len()).map(|p| format!("${p}")).join(","),
+        escape::escape_identifier(&table.name)
     );
 
-    let delete_stmt = client
-        .prepare(&delete_stmt)
-        .await
-        .context("internal error: preparing delete statement")?;
-    let insert_stmt = client
-        .prepare(&insert_stmt)
-        .await
-        .context("internal error: preparing insert statement")?;
-
-    for path in replace_files {
-        let file = load_file(file_config, path)
-            .await
-            .with_context(|| format!("loading replace file {path}"))?;
-        let reader = csv_async::AsyncReaderBuilder::new().create_reader(file);
-
-        replace_file(
-            file_config,
-            &key,
-            client,
-            &delete_stmt,
-            &insert_stmt,
-            reader,
-        )
-        .await
-        .with_context(|| format!("handling replace file {path}"))?;
-    }
-
-    Ok(())
-}
+    // First delete all of the matching rows.
+    let matching_cols = columns.iter().filter(|col| col.is_primary);
+    let delete_stmt = format!(
+        r#"
+        DELETE FROM {qualified_table_name}
+        WHERE ({cols}) IN (
+            SELECT {cols}
+            FROM {qualified_temp_table_name}
+        )"#,
+        cols = matching_cols.map(|col| &col.name).join(","),
+    );
+    let rows_changed = client.execute(&delete_stmt, &[]).await?;
+    tracing::info!(rows_changed, "deleted rows from {qualified_table_name}");
```

Comment:
What happens in we crash right here? Does the `INSERT` get executed after we restart?

Reply:
If we crash here then Fivetran will re-issue the command to us. We will re-run the `DELETE`, but this is okay because the relevant rows will have already been deleted, and then we'll run the `INSERT`. This leaves the user's data in an inconsistent state, but it's a tradeoff we're okay with right now